talk/basicconcepts/assert.tex

```diff
@@ -0,0 +1,89 @@
+\subsection[assert]{Assertions}
+
+\begin{frame}[fragile]
+  \frametitlecpp[98]{Assertions}
+  \begin{block}{Checking invariants in a program}
+    \begin{itemize}
+      \item An invariant is a property of a program that is always true, and the program might crash or yield wrong results if violated
+      \begin{itemize}
+        \item ``Here, `a' should always be positive''
+      \end{itemize}
+      \item This can be checked using \mintinline{cpp}{assert}
+      \item The program will be aborted if the assertion fails
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{assert in practice}
+    \begin{cppcode*}{}
+      #include <cassert>
+      double f(double a) {
+        // [...] do stuff with a
+        // [...] that should leave it positive
+        assert(a > 0.);
+        return std::sqrt(a);
+      }
+    \end{cppcode*}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[98]{Assertions}
+  \begin{block}{Good practices}
+    \begin{itemize}
+      \item Assertions are for developers and debugging; not for users
+      \item Use them to check important invariants of your program
+      \item Handle user-facing errors with error messages/exceptions
+      \item Assertions can impact the speed of a program. Therefore:
+      \begin{itemize}
+        \item Assertions are disabled when the macro \mintinline{cpp}{NDEBUG} is defined
+        \item Make sure it is defined when you release code
+        \item Many build systems do this automatically
+      \end{itemize}
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{Disabling assertions}
+    \begin{overprint}
+    \onslide<1>
+    \begin{cppcode*}{}
+      #define NDEBUG
+      #include <cassert>
+      double f(double a) {
+        assert(a > 0.); // no effect
+        return std::sqrt(a);
+      }
+    \end{cppcode*}
+    \onslide<2>
+    Compile entire project with NDEBUG:\\
+    \texttt{g++ \textcolor{blue}{-DNDEBUG} -O2 -W[...] test.cpp -o test.exe}
+    \end{overprint}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[17]{Static Assert}
```

Comment:
We already have slides on `static_assert` in `talk/morelanguage/constexpr.tex` :) Please unify!

Reply:
I wrote a bit about the fact that constexpr powers static_assert, but I wonder if I shouldn't just remove the slide. Opinions?

Reply:
I think I would remove the `static_assert` slide in the constexpr section, and just add here that the condition must be a constant expression.

Reply:
OK, done. The fact that it has to be `constexpr` has been mentioned. @bernhardmgruber

Reply:
Great!